src/Middleware/WebSockets/src/HandshakeHelpers.cs

```diff
@@ -72,5 +75,136 @@ public static string CreateResponseKey(string requestKey)
+                        // parameter in the corresponding extension negotiation response to the
+                        // offer with a value equal to or smaller than the received value.
+                        options.ClientMaxWindowBits = Math.Min(clientMaxWindowBits ?? 15, options.ClientMaxWindowBits);
+
+                        // If a received extension negotiation offer doesn't have the
+                        // "client_max_window_bits" extension parameter, the corresponding
+                        // extension negotiation response to the offer MUST NOT include the
+                        // "client_max_window_bits" extension parameter.
+                        builder.Append("; ").Append(WebSocketDeflateConstants.ClientMaxWindowBits).Append('=')
+                            .Append(options.ClientMaxWindowBits.ToString(CultureInfo.InvariantCulture));
+                    }
+                    else if (value.StartsWith(WebSocketDeflateConstants.ServerMaxWindowBits))
+                    {
+                        hasServerMaxWindowBits = true;
+                        var serverMaxWindowBits = ParseWindowBits(value, WebSocketDeflateConstants.ServerMaxWindowBits);
+                        // 8 is a valid value according to the spec, but our zlib implementation does not support it
+                        if (serverMaxWindowBits == 8)
+                        {
+                            return false;
+                        }
+
+                        // https://tools.ietf.org/html/rfc7692#section-7.1.2.1
+                        // A server accepts an extension negotiation offer with this parameter
+                        // by including the "server_max_window_bits" extension parameter in the
+                        // extension negotiation response to send back to the client with the
+                        // same or smaller value as the offer.
+                        options.ServerMaxWindowBits = Math.Min(serverMaxWindowBits ?? 15, options.ServerMaxWindowBits);
+                    }
+
+                    static int? ParseWindowBits(ReadOnlySpan<char> value, string propertyName)
+                    {
+                        var startIndex = value.IndexOf('=');
+
+                        // parameters can be sent without a value by the client, we'll use the values set by the app developer or the default of 15
+                        if (startIndex < 0)
+                        {
+                            return null;
+                        }
+
+                        if (!int.TryParse(value[(startIndex + 1)..], NumberStyles.Integer, CultureInfo.InvariantCulture, out int windowBits) ||
+                            windowBits < 8 ||
+                            windowBits > 15)
+                        {
+                            throw new WebSocketException(WebSocketError.HeaderError, $"invalid {propertyName} used: {value[(startIndex + 1)..].ToString()}");
```

Comment:
Exceptions caused by external user data... Probably better to ignore it (maybe log it), and return false.